nemo_text_processing/g2p/modules.py

```diff
@@ -404,39 +478,54 @@ def _normalize_dict(self, phoneme_dict_obj: dict) -> Tuple[DefaultDict[str, List
         g2p_dict = defaultdict(list)
         symbols = set()
         for word, prons in phoneme_dict_obj.items():
+            # process word
             # update word cases.
-            if self.set_graphemes_upper:
-                word_new = word.upper()
-            else:
-                word_new = word.lower()
+            word_new = set_grapheme_case(word, case=self.grapheme_case)
 
             # add grapheme symbols if `use_chars=True`.
             if self.use_chars:
-                # remove punctuations after a word.
+                # remove punctuations within a word. Punctuations can exist at the start, middle, and end of a word.
+                # TODO @xueyang: is the removal of a hyphen within a word expected, e.g. up-to-date -> uptodate?
                 word_no_punct = self.PUNCT_REGEX.sub('', word_new)
-                symbols.update(word_no_punct)
 
+                # add prefix to distinguish graphemes from phonemes.
+                symbols.update(self._prepend_prefix_for_one_word(word_no_punct))
```

Comment:
Do we want to prepend the grapheme symbol here? This would just add it to the symbols dictionary, which I'm not sure we want to do if we want to count [# + grapheme] as one token post-processing, and not as user input with # as punctuation.

Reply:
yes, IIUC, this is the right place to prepend a special symbol. `# + grapheme` is actually one token as the key in `self._token2id` dict. We don't use the predefined symbol set by the user, instead, we automatically summarize the grapheme and phoneme set from the dictionary. Without prepending the special symbol, it would not distinguish the token that overlaps grapheme and phoneme set.

Reply:
Can you confirm that it adds each `# + grapheme` as one symbol? Or does it add `#` separately from the grapheme?

Reply:
yes. the unit test explains: 
```
tests/nemo_text_processing/g2p/test_modules.py::TestIPAG2P::test_normalize_dict_with_graphemes_and_phonemes PASSED
```

Reply:
and your comments remind me of adding a validation func that avoid `grapheme_prefix` overlaps with character and punctuation lists.

Reply:
Got it, sounds good to me then! And thank you for adding the tests :)